README.md
# KtIrc

[![Build Status](https://travis-ci.org/csmith/KtIrc.svg?branch=master)](https://travis-ci.org/csmith/KtIrc)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/c01221cbf9cf413ba4d94cb8c80e334a)](https://www.codacy.com/app/csmith/KtIrc?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=csmith/KtIrc&amp;utm_campaign=Badge_Grade)
[![codecov](https://codecov.io/gh/csmith/KtIrc/branch/master/graph/badge.svg)](https://codecov.io/gh/csmith/KtIrc)
[![Download](https://api.bintray.com/packages/dmdirc/releases/ktirc/images/download.svg)](https://bintray.com/dmdirc/releases/ktirc/_latestVersion)

KtIrc is a Kotlin JVM library for connecting to and interacting with IRC servers.
It is still in an early stage of development.

## Features

#### Built for Kotlin

KtIrc is written in and designed for use in Kotlin; it uses extension methods,
DSLs, sealed classes, and so on, to make it much easier to use than an
equivalent Java library.

#### Coroutine-powered

KtIrc uses co-routines for all of its input/output which lets it deal with
IRC messages in the background while your app does other things, without
the overhead of creating a new thread per IRC client.

#### Modern IRC standards

KtIrc supports many IRCv3 features such as SASL authentication, message IDs,
server timestamps, replies, reactions, account tags, and more. These features
(where server support is available) make it easier to develop bots and
clients, and enhance IRC with new user-facing functionality.

## Setup

KtIrc is published to JCenter, so adding it to a gradle build is as simple as:

```groovy
repositories {
    jcenter()
}

dependencies {
    implementation("com.dmdirc:ktirc:<VERSION>")
}
```

## Usage

Clients are created using a DSL and the `IrcClient` function. At a minimum
you must specify a server and a profile. A simple bot might look like:

```kotlin
val client = IrcClient {
    server {
        host = "my.server.com"
    } 
    profile {
        nickname = "nick"
        username = "username"
        realName = "Hi there"
    }
}

client.onEvent { event ->
    when (event) {
        is ServerReady ->
            client.sendJoin("#ktirc")
        is ServerDisconnected ->
            client.connect()
        is MessageReceived ->
            if (event.message == "!test")
                client.reply(event, "Test successful!")
    }
}

client.connect()
```

## Developing KtIrc

### Lifecycle of a message

![architecture diagram](docs/sequence.png) 

The `LineBufferedSocket` class receives bytes from the IRC server. Whenever it
encounters a complete line (terminated by a `CR`, `LF` or `CRLF`), it passes it
to the `IrcClient` as a `ByteArray`. The `MessageParser` breaks up the line
into its component parts (tags, prefixes, commands, and parameters) and returns
them as an `IrcMessage`.
 
The `IrcMessage` is given to the `MessageHandler`, which tries to find a
processor that can handle the command in the message. The processor's job is
to convert the message into an `IrcEvent` subclass. Processors do not get
given any contextual information or state, their job is simply to convert
the message as received into an event.

The events are returned to the `MessageHandler` which then passes them on
to all registered event handlers. The job of the event handlers is twofold:
firstly, use the events to update the state of KtIrc (for example, after
receiving a `JOIN` message, the `ChannelStateHandler` will add the user
to the list of users in the channel, while the `UserStateHandler` may update
the user's hostname if we hadn't previously seen it). Secondly, the event
handlers may themselves raise events. This is useful for higher-order
events such as `ServerReady` that depend on a variety of factors and
states.

Handlers themselves may not keep state, as they will be shared across
multiple instances of `IrcClient` and won't be reset on reconnection.
State is instead stored in the various `*State` properties of the
`IrcClient` such as `serverState` and `channelState`. Fields that
should not be exposed to users of KtIrc can be placed in these
public state objects but marked as `internal`.

All the generated events (from processors or from event handlers) are
passed to the `IrcClient`, which in turn passes them to the library
user via the delegates passed to the `onEvent` method. 

### Contributing

Contributing is welcomed and encouraged! Please try to add unit tests for new features,
and maintain a code style consistent with the existing code.

### Licence

The code in this repository is released under the MIT licence. See the
[LICENCE](LICENCE) file for more info.


CHANGELOG
vNEXT (in development)

 * Away support:
   * Added sendAway() method
   * Added UserAway event and fanned-out ChannelAway
   * Away message is now updated in the UserState
 * Added IrcClient.localUser property, returning the local User
 * Deprecated ServerState.localNickname in favour of localUser.
 * (Internal) improved the way byte buffers are used to
   reduce array copying and clean up code

v1.0.1

 * Fixed issue with very long packets not fitting in buffers
   after TLS decryption.
 * Improved error message when IrcClient.connect() is called
   while already connected
 * Made IrcClient.disconnect() block until the disconnect has
   been processed. This will be a very short period of time,
   and allows callers to immediately call connect() again
   afterwards if they wish
 * Fixed various issues with disconnecting TLS connections
   not behaving correctly

v1.0.0

 * Replaced Ktor dependency with custom socket handling, which fixes
   fatal issue when connecting to servers over TLS that request a
   client certificate.
 * Added NicknameChangeRequired event for the case when a nickname is
   not allowed during connection and *MUST* be changed
 * BREAKING: Added 'PreferIPv6' behaviour config, defaulting to true.
   With this configuration KtIrc will try to use IPv6 if the server
   publishes AAAA records.
 * BREAKING: Default port is now 6697, TLS is enabled by default

v0.11.0

 * Added SourcedEvent interface for events that have a user source attached
 * Added ChannelJoinFailed event
 * Added ChannelMembershipAdjustment interface for events that concern
   users joining/leaving channels
 * Added missing user parameter to ModeChangedEvent

v0.10.3

 * Fix issue parsing CTCPs when the content contained multi-byte chars
 * Fixed multiple issues when receiving malformed input from the server:
   * Invalid server-time tags are now ignored and the current time used
   * PRIVMSGs, NOTICEs and TOPICs with no target/message are now ignored
   * CAPs with invalid or missing arguments are now ignored
   * Better handling for duplicate JOIN messages
   * Ignored messages are now logged more consistently

v0.10.2

 * Fix handling of multiple long lines sometimes dropping part of the line
 * Support lines with 8191 bytes of tags, as allowed by the message-tags spec
 * Fix a few more instances of occasional ConcurrentModificationExceptions

v0.10.1

 * Added NickChangeFailed event for when nicknames are in use, banned, etc
 * Added sendPart method
 * Fix occasional ConcurrentModificationException when adding state
 * (Internal) Moved message processors into their own package
 * (Internal) Added ExperimentalIrcClient interface for features not yet ready for release

v0.10.0

 * Batch start and end events are no longer included in BatchReceived events
 * Batches now expose complete metadata from their start event
 * Added support for labelled-replies capability and label message tags
 * Added support for message-tags capability after it moved from draft
 * Added new methods for sending raw lines to the IRC server
   * send(tagMap(...), command, arguments) replaces send(line)
   * send(command, arguments) is available if no tags are to be sent
   * the line is built automatically (with spaces/' :' added appropriately)
   * send(line) is deprecated and will be removed after v1.0.0
 * Add 'alwaysEchoMessages' behaviour that makes it easier to deal with servers
   that don't support the echo message capability
 * Added TargetedEvent, a subclass of IrcEvent, for all events that are
   targeted to a user or channel
 * (Internal) Added annotation to track removal of deprecated methods
 * (Internal) Migrate unit tests to use Mockk instead of Mockito

v0.9.0

 * Improve DSL for creating an IrcClient to allow parameters to be passed to server and profile
   e.g. IrcClient { server("irc.example.com", 6667) }
 * Add behaviour options
   * requestModesOnJoin - automatically sends a MODE request when joining a channel
 * Events now have a `metadata` property instead of a `time` (and time is available in metadata)
   * IrcEvent.time is now deprecated but will remain until after v1.0.0.
   * Metadata now contains the message ID, if any.
     * ActionReceived.messageId and MessageReceived.messageId are now deprecated, to be removed after v1.0.0.
   * Metadata now contains the event's batch ID, if any.
 * Added support for batches
   * All events in a batch are buffered until the batch is finished
   * The events are then published together in a single `BatchReceived` event
 * Added support for CHGHOST messages
 * (Internal) Improve performance when the MessageHandler is finding a processor for a message
 * (Internal) Introduced event mutators
   * Event mutators are now responsible for handling changing events in response to state
     e.g. ChannelFanOutMutator creates Channel* events for global quits/nick changes/etc
   * Event handlers now just handle events, and don't return anything

v0.8.0

 * Added support for SCRAM-SHA-1 and SCRAM-SHA-256 SASL mechanisms
 * Added MotdLineReceived event
 * Added topic events and state
 * Add utility method IrcClient.isChannel(String) to identify if a target is a channel or not
 * (Internal) Move event handlers into their own package

v0.7.0

 * Fixed experimental API warnings when using IrcClient
 * BREAKING: IrcClients are now constructed using a DSL
   * Users of the library no longer need to care about the implementing class
   * Facilitates adding more options in the future without breaking existing implementations
 * SASL improvements
   * The enabled mechanisms can now be configured (in the SASL DSL)
   * Added support for EXTERNAL mechanism, disabled by default
   * Now attempts to renegotiate if the server doesn't recognise the SASL mechanism that was tried
 * Added UserNickChanged and corresponding ChannelNickChanged events

src/main/kotlin/com/dmdirc/ktirc/Dsl.kt
package com.dmdirc.ktirc

/**
 * Dsl marker for [IrcClient] dsl.
 */
@DslMarker
annotation class IrcClientDsl

internal data class IrcClientConfig(
        val server: ServerConfig,
        val profile: ProfileConfig,
        val behaviour: ClientBehaviour,
        val sasl: SaslConfig?)

/**
 * Dsl for configuring an IRC Client.
 *
 * [server] and [profile] blocks are required. The full range of configuration options are:
 *
 * ```
 * server {
 *     host = "irc.example.com"     // Required
 *     port = 6667
 *     useTls = true
 *     password = "H4ckTh3Pl4n3t"
 * }
 *
 * profile {
 *     nickname = "MyBot"           // Required
 *     username = "bot"
 *     realName = "Botomatic v1.2"
 * }
 *
 * behaviour {
 *     requestModesOnJoin = true
 *     alwaysEchoMessages = true
 *     preferIPv6 = false
 * }
 *
 * sasl {
 *     mechanisms += "PLAIN" // or to set the list from scratch:
 *     mechanisms("PLAIN")
 *
 *     username = "botaccount"
 *     password = "s3cur3"
 * }
 * ```
 */
@IrcClientDsl
class IrcClientConfigBuilder {

    private var server: ServerConfig? = null
        set(value) {
            check(field == null) { "server may only be specified once" }
            check(!value?.host.isNullOrEmpty()) { "server.host must be specified" }
            field = value
        }

    private var profile: ProfileConfig? = null
        set(value) {
            check(field == null) { "profile may only be specified once" }
            check(!value?.nickname.isNullOrEmpty()) { "profile.nickname must be specified" }
            field = value
        }

    private var behaviour: BehaviourConfig? = null
        set(value) {
            check(field == null) { "behaviour may only be specified once" }
            field = value
        }

    private var sasl: SaslConfig? = null
        set(value) {
            check(field == null) { "sasl may only be specified once" }
            field = value
        }

    /**
     * Configures the server that the IrcClient will connect to.
     *
     * See [ServerConfig] for details of each parameter.
     *
     * @param block Optional additional configuration to apply to the [ServerConfig]
     */
    @IrcClientDsl
    fun server(host: String? = null, port: Int? = null, useTls: Boolean? = null, password: String? = null, block: (ServerConfig.() -> Unit)? = null) {
        server = ServerConfig().apply {
            host?.let { this.host = it }
            port?.let { this.port = it }
            useTls?.let { this.useTls = it }
            password?.let { this.password = it }
            block?.let { apply(it) }
        }
    }

    /**
     * Configures the profile of the IrcClient user.
     *
     * See [ProfileConfig] for details of each parameter.
     *
     * @param block Optional additional configuration to apply to the [ProfileConfig]
     */
    @IrcClientDsl
    fun profile(nickname: String? = null, username: String? = null, realName: String? = null, block: (ProfileConfig.() -> Unit)? = null) {
        profile = ProfileConfig().apply {
            nickname?.let { this.nickname = it }
            username?.let { this.username = it }
            realName?.let { this.realName = it }
            block?.let { apply(it) }
        }
    }

    /**
     * Configures the behaviour of the client (optional).
     */
    @IrcClientDsl
    fun behaviour(block: BehaviourConfig.() -> Unit) {
        behaviour = BehaviourConfig().apply(block)
    }

    /**
     * Configures SASL authentication (optional).
     */
    @IrcClientDsl
    fun sasl(block: SaslConfig.() -> Unit) {
        sasl = SaslConfig().apply(block)
    }

    internal fun build() =
            IrcClientConfig(
                    checkNotNull(server) { "Server must be specified " },
                    checkNotNull(profile) { "Profile must be specified" },
                    behaviour ?: BehaviourConfig(),
                    sasl)

}

/**
 * Dsl for configuring a server.
 */
@IrcClientDsl
class ServerConfig {
    /** The hostname (or IP address) of the server to connect to. */
    var host: String = ""
    /** The port to connect on. Defaults to 6697. */
    var port: Int = 6697
    /** Whether or not to use TLS (an encrypted connection). */
    var useTls: Boolean = true
    /** The password required to connect to the server, if any. */
    var password: String? = null
}

/**
 * Dsl for configuring a profile.
 */
@IrcClientDsl
class ProfileConfig {
    /** The initial nickname to use when connecting. */
    var nickname: String = ""
    /** The username (used in place of an ident response) to provide to the server. */
    var username: String = "KtIrc"
    /** The "real name" to provide to the server. */
    var realName: String = "KtIrc User"
}

/**
 * Dsl for configuring SASL authentication.
 *
 * By default the `PLAIN`, `SCRAM-SHA-1`, and `SCRAM-SHA-256` methods will be enabled if SASL is configured.
 *
 * You can modify the mechanisms either by editing the [mechanisms] collection:
 *
 * ```
 * mechanisms += "EXTERNAL"
 * mechanisms.remove("PLAIN")
 * ```
 *
 * or by calling the [mechanisms] function with all the mechanisms you wish
 * to enable:
 *
 * ```
 * mechanisms("PLAIN", "EXTERNAL")
 * ```
 *
 * Priority of mechanisms is determined by KtIrc, regardless of the order
 * they are specified in here.
 */
@IrcClientDsl
class SaslConfig {
    /** The SASL mechanisms to enable. */
    val mechanisms: MutableCollection<String> = mutableSetOf("PLAIN", "SCRAM-SHA-1", "SCRAM-SHA-256")
    /** The username to provide when authenticating using SASL. */
    var username: String = ""
    /** The username to provide when authenticating using SASL. */
    var password: String = ""

    /**
     * Replaces all enabled SASL mechanisms with the given ones.
     */
    @IrcClientDsl
    fun mechanisms(vararg methods: String) {
        with(this.mechanisms) {
            clear()
            addAll(methods)
        }
    }
}

/**
 * Dsl for configuring the behaviour of an [IrcClient].
 */
@IrcClientDsl
class BehaviourConfig : ClientBehaviour {
    override var requestModesOnJoin = false
    override var alwaysEchoMessages = false
    override var preferIPv6 = true
}
